README.md
# ESLint Config

This ESLint configuration targets plain JavaScript, and is intended to be complemented with additional rulesets for suit projects using TypeScript, React, unit tests, etc.

## Usage

A quick note, this ruleset is incapable of installing its own plugins due to [limitations](https://github.com/eslint/eslint/issues/3458) within current versions of ESLint. Any requirements will be marked as peer dependencies.

### 1. Install

```sh
npm i --save-dev @silicon-soldier/eslint-config
```

### 2. Configure

Create an ESLint config file called `.eslintrc.json` (or whichever extension you need/prefer).

```json
{
    "extends": "@silicon-soldier/eslint-config",
    "parserOptions": {
        "ecmaVersion": 2015
    }
}
```

Some things to note are;

- ECMA version is specified to ensure ESLint interprets source according to the correct spec version.

### 3. Integrate

ESLint isn't magic. To actually get value out, you'll need to integrate it with your projects existing tooling to maximise value. Some other recommendations are;

- Integrating with Git Hooks using [Husky](https://www.npmjs.com/package/husky), good for avoiding accidental commits of unlinted code.
- Including it in integration tests so that bad linting can be caught for every push/PR in a reliable fashion.
- Installing plugins for your IDE that will automatically lint what is being viewed, and overlay issues.

## Roadmap

This roadmap will eventually be relocated to the wiki.

### 1.0.0

- [x] Improve dog-fooding experience.
- [ ] Improve ruleset tests with pass and fail scenarios to permit catching of regressions and breaks from existing style.
- [ ] Document code style.
- [ ] Implement CI.

### 2.0.0

- [ ] Prefer JSX syntax for XML derivatives when detected in a string.
- [ ] JSDoc enforcement.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0]

### Changed

- Sorting of imports is now case insensitive.

## [0.1.2] - 2019-04-14

### Fixed

- Misconfiguration in `.npmignore`.

## [0.1.1] - 2019-04-13

### Added

- Dog-feeding config against itself.

### Removed

- TypeScript to facilitate more effective dog-fooding.

## [0.1.0] - 2019-03-12

Initial developmental release.


tests/integration/sample.js
/**
 * A test function.
 * @param {string} param A string
 */
export function test(param)
{
    const theThing = "foo";

    return `${theThing} - ${param}`;
}

/**
 * A test class.
 */
export class Foo
{
    constructor()
    {
        this.thing = test("bar");
    }
}


tests/integration/sort-imports/afile1.js
export const short = "foo";
export const longVarName = "bar";
export const medium = "foobar";


src/ruleset.js
/* eslint-disable id-length */
export default {
    rules: {

        /**
         * Consistent and readable array usage.
         */
        "array-bracket-newline": [
            "error",
            "consistent",
        ],

        /**
         * @todo
         */
        "array-bracket-spacing": [
            "error",
            "always",
            { arraysInArrays: false, },
        ],

        /**
         * Consistent and readable array usage.
         */
        "array-element-newline": [
            "error",
            "consistent",
        ],

        /**
         * Concise.
         */
        "arrow-body-style": [
            "error",
            "as-needed",
        ],

        /**
         * Concise.
         */
        "arrow-parens": [
            "error",
            "as-needed",
            { requireForBlockBody: true, },
        ],

        /**
         * @todo
         */
        "arrow-spacing": "error",

        /**
         * Prevent unnecessary whitespace.
         * Spacing is managed via a pre-comment empty line and bracket
         * placement.
         */
        "block-spacing": "error",

        /**
         * Spaces out code in a reliable and readable fashion.
         * Easier to locate blocks.
         */
        "brace-style": [
            "error",
            "allman",
        ],

        /**
         * Consistent variable structuring without sacrificing space.
         */
        camelcase: [
            "error",
        ],

        /**
         * Quicker source manipulations.
         */
        "comma-dangle": [
            "error",
            "always",
        ],

        /**
         * Prevent clutter.
         */
        "comma-spacing": "error",

        /**
         * @todo
         */
        "comma-style": "error",

        /**
         * @todo
         */
        "computed-property-spacing": "error",

        /**
         * Eliminate cognetive load associated with figuring out what `this` is.
         */
        "consistent-this": [
            "error",
            "that",
        ],

        /**
         * @todo
         */
        "constructor-super": "error",

        /**
         * Improve compatibilty across tools.
         */
        "eol-last": "error",

        /**
         * @todo
         */
        "func-call-spacing": "error",

        /**
         * No redundant idenifiers.
         */
        "func-names": [
            "error",
            "as-needed",
        ],

        /**
         * @todo
         */
        "func-style": [
            "error",
            "declaration",
            { allowArrowFunctions: true, },
        ],

        /**
         * Consistent function paramater placement.
         */
        "function-paren-newline": [
            "error",
            "consistent",
        ],

        /**
         * Consistent generator functions.
         */
        "generator-star-spacing": [
            "error",
            {
                after: true,
                anonymous: "after",
                before: false,
                method: "before",
            },
        ],

        /**
         * Keep identifiers short to reduce source verbosity and encourage
         * smarter structuring of data.
         */
        "id-length": [
            "error",
            { max: 12, },
        ],

        /**
         * @todo
         */
        "implicit-arrow-linebreak": "error",

        /**
         * Consistent indentation across all editors.
         */
        indent: [
            "error",
            4,
        ],

        /**
         * Match XML markup and other code style.
         */
        "jsx-quotes": [
            "error",
            "prefer-double",
        ],

        /**
         * @todo
         */
        "key-spacing": "error",

        /**
         * @todo
         */
        "keyword-spacing": "error",

        /**
         * @todo
         */
        "line-comment-position": "error",

        /**
         * @todo
         */
        "linebreak-style": "error",

        /**
         * Frame comments to make them easier to pick out.
         */
        "lines-around-comment": [
            "error",
            {
                allowBlockEnd: true,
                allowBlockStart: true,
                beforeLineComment: true,

            },
        ],

        /**
         * @todo
         */
        "lines-between-class-members": "error",
        "max-depth": [
            "error",
            { max: 6, },
        ],

        /**
         * Keep source concise, and prevent dependence on line wrapping.
         * Also permits a more pleasent experience reviewing code on various
         * devices.
         */
        "max-len": [
            "error",
            {

                /**
                 * Keep source concise and readable.
                 */
                code: 100,

                /**
                 * Keep comments concise and readable, forming an "island"
                 * within the source via padding to better distinguish.
                 */
                comments: 80,
            },
        ],

        /**
         * Keep files at a maintainable size. Comments are ignored which
         * translates to significant room for well documented logic.
         * @todo Is syntax like a bracket included in the calculation?
         */
        "max-lines": [
            "warn",
            {
                max: 370,
                skipBlankLines: true,
                skipComments: true,
            },
        ],

        /**
         * Prevent large blocks of unreusable code from forming.
         */
        "max-lines-per-function": [
            "warn",
            {
                IIFEs: true,
                max: 70,
            },
        ],

        /**
         * Keep program flow easily traceable.
         */
        "max-nested-callbacks": [
            "error",
            { max: 1, },
        ],

        /**
         * @todo
         */
        "max-params": "error",

        /**
         * Avoid high density complex code.
         * @todo Need to check how this plays with function literals
         */
        "max-statements-per-line": "error",
        "multiline-comment-style": [
            "error",

            /**
             * @todo Shouldn't affect JSDoc
             */
            "separate-lines",
        ],

        /**
         * @todo
         */
        "new-cap": [
            "error",
            { capIsNew: true, },
        ],

        /**
         * @todo
         */
        "new-parens": "error",

        /**
         * @todo
         */
        "newline-per-chained-call": "error",

        /**
         * Consistent array declarations.
         */
        "no-array-constructor": "error",

        /**
         * Avoid hard to understand JS bitwise logic. Usage is usually by
         * mistake, overrides should be used if use is deliberate.
         */
        "no-bitwise": "error",

        /**
         * @todo
         */
        "no-class-assign": "error",

        /**
         * @todo Makes control flow MUCH harder to follow
         */
        "no-confusing-arrow": [
            "error",
            { allowParens: false, },
        ],

        /**
         * @todo
         */
        "no-const-assign": "error",

        /**
         * Continue statements interrupt normal flow and as loop blocks are not
         * commonly trated as function blocks, bad states may be created.
         */
        "no-continue": "error",

        /**
         * @todo
         */
        "no-dupe-class-members": "error",

        /**
         * @todo
         */
        "no-duplicate-imports": [
            "error",
            { includeExports: true, },
        ],

        /**
         * @todo
         */
        "no-extra-parens": [
            "error",
            "all",
            { nestedBinaryExpressions: false, },
        ],

        /**
         * @todo
         */
        "no-inline-comments": "error",

        /**
         * @todo
         */
        "no-lonely-if": "error",

        /**
         * @todo
         */
        "no-mixed-operators": [
            "error",
            { allowSamePrecedence: true, },
        ],

        /**
         * Force consistent indentation means.
         */
        "no-mixed-spaces-and-tabs": "error",

        /**
         * @todo
         */
        "no-multi-assign": "error",

        /**
         * Prevent excess useless whitespace.
         */
        "no-multiple-empty-lines": "error",

        /**
         * @todo
         */
        "no-nested-ternary": "error",

        /**
         * @todo
         */
        "no-new-object": "error",

        /**
         * @todo
         */
        "no-new-symbol": "error",

        /**
         * @todo
         */
        "no-restricted-syntax": [
            "error",
            "WithStatement",
        ],

        /**
         * @todo
         */
        "no-tabs": "error",

        /**
         * Prevent usage of class resources before parent has been constructed.
         */
        "no-this-before-super": "error",

        /**
         * Prevent garbage whitespace at the end of lines.
         */
        "no-trailing-spaces": "error",

        /**
         * @todo
         */
        "no-underscore-dangle": "error",

        /**
         * @todo
         */
        "no-unneeded-ternary": "error",

        /**
         * @todo
         */
        "no-useless-computed-key": "error",

        /**
         * @todo
         */
        "no-useless-constructor": "error",

        /**
         * @todo
         */
        "no-useless-rename": "error",

        /**
         * Prevent usage of loosly scoped `var`.
         */
        "no-var": "error",

        /**
         * @todo
         */
        "no-whitespace-before-property": "error",

        /**
         * @todo
         */
        "nonblock-statement-body-position": "error",

        /**
         * @todo
         */
        "object-curly-newline": [
            "error",
            {
                minProperties: 2,
                multiline: true,
            },
        ],

        /**
         * @todo
         */
        "object-curly-spacing": [
            "error",
            "always",
        ],

        /**
         * @todo
         */
        "object-property-newline": [
            "error",
            { allowMultiplePropertiesPerLine: false, },
        ],

        /**
         * @todo
         */
        "object-shorthand": [
            "error",
            "always",
            {
                avoidExplicitReturnArrows: true,
                avoidQuotes: true,
                ignoreConstructors: true,
            },
        ],

        /**
         * @todo
         */
        "one-var": [
            "error",
            "never",
        ],

        /**
         *
         */
        "one-var-declaration-per-line": [
            "error",
            "always",
        ],

        /**
         * @todo
         */
        "operator-assignment": "error",

        /**
         * @todo
         */
        "operator-linebreak": [
            "error",
            "before",
        ],

        /**
         * @todo
         */
        "padded-blocks": [
            "error",
            "never",
            { allowSingleLineBlocks: false, },
        ],

        /**
         * @todo
         */
        "prefer-arrow-callback": [
            "error",
            {
                allowNamedFunctions: true,
                allowUnboundThis: false,
            },
        ],

        /**
         * @todo
         */
        "prefer-const": "error",

        /**
         * @todo
         */
        "prefer-numeric-literals": "error",

        /**
         * @todo
         */
        "prefer-object-spread": "error",

        /**
         * @todo
         */
        "prefer-rest-params": "error",

        /**
         * @todo
         */
        "prefer-spread": "error",

        /**
         * Template strings are easier to read compared to a lengthy
         * concatenation sequence.
         */
        "prefer-template": "error",

        /**
         * @todo This should be a JSX specific rule.
         */
        "quote-props": [
            "error",
            "as-needed",
        ],

        /**
         * An apostrophe is highly likely to appear in strings. This also
         * maintains consistency with JSX string preferences throughout the
         * community.
         *
         * Beyond that, this is consistent with strictly typed languages like
         * C#.
         */
        quotes: [
            "error",
            "double",
            {
                allowTemplateLiterals: true,
                avoidEscape: true,
            },
        ],

        /**
         * @todo
         */
        "require-yield": "error",

        /**
         * @todo
         */
        "rest-spread-spacing": "error",

        /**
         * Prevent dependence on poorly understood and easily misused magic
         * statement termination.
         */
        semi: "error",

        /**
         * @todo
         */
        "semi-style": "error",

        /**
         * Easier time inspecting imports. Less wars over ordering.
         */
        "sort-imports": [
            "error",
            { ignoreCase: true, },
        ],

        /**
         * @todo
         */
        "sort-keys": [
            "error",
            "asc",
            {
                caseSensitive: false,
                natural: true,
            },
        ],

        /**
         * @todo
         */
        "space-before-blocks": "error",

        /**
         * @todo
         */
        "space-before-function-paren": [
            "error",
            {
                anonymous: "always",
                asyncArrow: "never",
                named: "never",
            },
        ],

        /**
         * @todo
         */
        "space-in-parens": [
            "error",
            "never",
        ],

        /**
         * @todo
         */
        "space-infix-ops": "error",

        /**
         * @todo
         */
        "space-unary-ops": "error",

        /**
         * @todo
         */
        "spaced-comment": [
            "error",
            "always",
            {
                markers: [
                    "!",
                ],
            },
        ],

        /**
         * @todo
         */
        "switch-colon-spacing": "error",

        /**
         * @todo
         */
        "symbol-description": "error",

        /**
         * @todo
         */
        "template-curly-spacing": [
            "error",
            "never",
        ],

        /**
         * @todo
         */
        "template-tag-spacing": "error",

        /**
         * @todo
         */
        "unicode-bom": "error",

        /**
         * @todo
         */
        "yield-star-spacing": [
            "error",
            "after",
        ],
    },
};


src/ruleset.ts
import { Linter } from "eslint";

const config: Linter.Config = {
    rules: {
        "array-bracket-newline": [
            "error",
            "consistent"
        ],
        "array-bracket-spacing": [
            "error",
            "always",
            {
                arraysInArrays: false
            }
        ],
        "array-element-newline": [
            "error",
            "consistent"
        ],
        "arrow-body-style": [
            "error",
            "as-needed"
        ],
        "arrow-parens": [
            "error",
            "as-needed",
            {
                requireForBlockBody: true
            }
        ],
        "arrow-spacing": "error",
        "block-spacing": "error",
        // Inherently spaces out code in a more readable fashion.
        // Makes it easier to locate blocks.
        "brace-style": [
            "error",
            "allman"
        ],
        camelcase: [
            "error"
        ],
        "comma-dangle": [
            "error",
            "always"
        ],
        "comma-spacing": "error",
        "comma-style": "error",
        "computed-property-spacing": "error",
        "consistent-this": [
            "error",
            "that"
        ],
        "constructor-super": "error",
        "eol-last": "error",
        "func-call-spacing": "error",
        "func-names": [
            "error",
            "as-needed"
        ],
        "func-style": [
            "error",
            "declaration",
            {
                allowArrowFunctions: true
            }
        ],
        "function-paren-newline": [
            "error",
            "consistent"
        ],
        "generator-star-spacing": [
            "error",
            {
                before: false,
                after: true,
                anonymous: "after",
                method: "before"
            }
        ],
        "id-length": [
            "error",
            {
                max: 12
            }
        ],
        "implicit-arrow-linebreak": "error",
        "indent": [
            "error",
            4
        ],
        "jsx-quotes": [
            "error",
            "prefer-double"
        ],
        "key-spacing": "error",
        "keyword-spacing": "error",
        "line-comment-position": "error",
        "linebreak-style": "error",
        "lines-around-comment": [
            "error",
            {
                beforeLineComment: true,
                allowBlockStart: true,
                allowBlockEnd: true,

            }
        ],
        "lines-between-class-members": "error",
        "max-depth": [
            "error",
            {
                max: 6
            }
        ],
        "max-len": [
            "error",
            {
                code: 100,
                comments: 80
            }
        ],
        "max-lines": [
            "warn",
            {
                max: 370,
                skipBlankLines: true,
                skipComments: true
            }
        ],
        "max-lines-per-function": [
            "warn",
            {
                max: 70,
                IIFEs: true
            }
        ],
        "max-nested-callbacks": [
            "error",
            {
                max: 1
            }
        ],
        "max-params": "error",
        // Need to check how this plays with function literals
        "max-statements-per-line": "error",
        "multiline-comment-style": [
            "error",
            "separate-lines"// Shouldn't affect JSDoc
        ],
        "new-cap": [
            "error",
            {
                "capIsNew": true,
            }
        ],
        "new-parens": "error",
        "newline-per-chained-call": "error",
        "no-array-constructor": "error",
        // 99% of the time, it'll be a mistake.
        // Override at location to indicate deliberate use.
        "no-bitwise": "error",
        // Makes control flow MUCH harder to follow
        "no-confusing-arrow": [
            "error",
            {
                allowParens: false
            }
        ],
        "no-const-assign": "error",
        "no-continue": "error",
        "no-class-assign": "error",
        "no-dupe-class-members": "error",
        "no-duplicate-imports": [
            "error",
            {
                includeExports: true
            }
        ],
        "no-extra-parens": [
            "error",
            "all",
            {
                nestedBinaryExpressions: false
            }
        ],
        "no-inline-comments": "error",
        "no-lonely-if": "error",
        "no-new-symbol": "error",
        "no-mixed-operators": [
            "error",
            {
                allowSamePrecedence: true
            }
        ],
        "no-mixed-spaces-and-tabs": "error",
        "no-multi-assign": "error",
        "no-multiple-empty-lines": "error",
        "no-nested-ternary": "error",
        "no-new-object": "error",
        "no-restricted-syntax": [
            "error",
            "WithStatement"
        ],
        "no-tabs": "error",
        "no-trailing-spaces": "error",
        "no-this-before-super": "error",
        "no-underscore-dangle": "error",
        "no-unneeded-ternary": "error",
        "no-useless-computed-key": "error",
        "no-useless-constructor": "error",
        "no-useless-rename": "error",
        "no-var": "error",
        "no-whitespace-before-property": "error",
        "nonblock-statement-body-position": "error",
        "object-curly-newline": [
            "error",
            {
                multiline: true,
                minProperties: 2
            }
        ],
        "object-curly-spacing": [
            "error",
            "always"
        ],
        "object-property-newline": [
            "error",
            {
                "allowMultiplePropertiesPerLine": false
            }
        ],
        "object-shorthand": [
            "error",
            "always",
            {
                avoidQuotes: true,
                ignoreConstructors: true,
                avoidExplicitReturnArrows: true
            }
        ],
        "one-var": [
            "error",
            "never"
        ],
        "one-var-declaration-per-line": [
            "error",
            'always'
        ],
        "operator-assignment": "error",
        "operator-linebreak": [
            "error",
            "before"
        ],
        "padded-blocks": [
            "error",
            "never",
            {
                allowSingleLineBlocks: false
            }
        ],
        "prefer-arrow-callback": [
            "error",
            {
                allowNamedFunctions: true,
                allowUnboundThis: false
            }
        ],
        "prefer-const": "error",
        "prefer-numeric-literals": "error",
        "prefer-object-spread": "error",
        "prefer-rest-params": "error",
        "prefer-spread": "error",
        "prefer-template": "error",
        "quote-props": [
            "error",
            "as-needed"
        ],
        quotes: [
            "error",
            "double",
            {
                avoidEscape: true,
                allowTemplateLiterals: true
            }
        ],
        "require-yield": "error",
        "rest-spread-spacing": "error",
        semi: "error",
        "spaced-comment": [
            "error",
            "always",
            {
                markers: [
                    "!"
                ]
            }
        ],
        "semi-style": "error",
        "sort-keys": [
            "error",
            "asc",
            {
                caseSensitive: false,
                natural: true
            }
        ],
        "sort-imports": "error",
        "space-before-blocks": "error",
        "space-before-function-paren": [
            "error",
            {
                anonymous: "always",
                named: "never",
                asyncArrow: "never"
            }
        ],
        "space-in-parens": [
            "error",
            "never"
        ],
        "space-infix-ops": "error",
        "space-unary-ops": "error",
        "switch-colon-spacing": "error",
        "symbol-description": "error",
        "template-curly-spacing": [
            "error",
            "never"
        ],
        "template-tag-spacing": "error",
        "unicode-bom": "error",
        "yield-star-spacing": [
            "error",
            "after"
        ]
    }
};

export default config;


src/eslint.ts
import { Linter } from "eslint";

const config: Linter.Config = {
    rules: {
        "array-bracket-newline": [
            "error",
            "consistent"
        ],
        "array-bracket-spacing": [
            "error",
            "always",
            {
                arraysInArrays: false
            }
        ],
        "array-element-newline": [
            "error",
            "consistent"
        ],
        "arrow-body-style": [
            "error",
            "as-needed"
        ],
        "arrow-parens": [
            "error",
            "as-needed",
            {
                requireForBlockBody: true
            }
        ],
        "arrow-spacing": "error",
        "block-spacing": "error",
        // Inherently spaces out code in a more readable fashion.
        // Makes it easier to locate blocks.
        "brace-style": [
            "error",
            "allman"
        ],
        camelcase: [
            "error"
        ],
        "comma-dangle": [
            "error",
            "always"
        ],
        "comma-spacing": "error",
        "comma-style": "error",
        "computed-property-spacing": "error",
        "consistent-this": [
            "error",
            "that"
        ],
        "constructor-super": "error",
        "eol-last": "error",
        "func-call-spacing": "error",
        "func-names": [
            "error",
            "as-needed"
        ],
        "func-style": [
            "error",
            "declaration",
            {
                allowArrowFunctions: true
            }
        ],
        "function-paren-newline": [
            "error",
            "consistent"
        ],
        "generator-star-spacing": [
            "error",
            {
                before: false,
                after: true,
                anonymous: "after",
                method: "before"
            }
        ],
        "id-length": [
            "error",
            {
                max: 12
            }
        ],
        "implicit-arrow-linebreak": "error",
        "indent": [
            "error",
            4
        ],
        "jsx-quotes": [
            "error",
            "prefer-double"
        ],
        "key-spacing": "error",
        "keyword-spacing": "error",
        "line-comment-position": "error",
        "linebreak-style": "error",
        "lines-around-comment": [
            "error",
            {
                beforeLineComment: true,
                allowBlockStart: true,
                allowBlockEnd: true,

            }
        ],
        "lines-between-class-members": "error",
        "max-depth": [
            "error",
            {
                max: 6
            }
        ],
        "max-len": [
            "error",
            {
                code: 100,
                comments: 80
            }
        ],
        "max-lines": [
            "warn",
            {
                max: 370,
                skipBlankLines: true,
                skipComments: true
            }
        ],
        "max-lines-per-function": [
            "warn",
            {
                max: 70,
                IIFEs: true
            }
        ],
        "max-nested-callbacks": [
            "error",
            {
                max: 1
            }
        ],
        "max-params": "error",
        // Need to check how this plays with function literals
        "max-statements-per-line": "error",
        "multiline-comment-style": [
            "error",
            "separate-lines"// Shouldn't affect JSDoc
        ],
        "new-cap": [
            "error",
            {
                "capIsNew": true,
            }
        ],
        "new-parens": "error",
        "newline-per-chained-call": "error",
        "no-array-constructor": "error",
        // 99% of the time, it'll be a mistake.
        // Override at location to indicate deliberate use.
        "no-bitwise": "error",
        // Makes control flow MUCH harder to follow
        "no-confusing-arrow": [
            "error",
            {
                allowParens: false
            }
        ],
        "no-const-assign": "error",
        "no-continue": "error",
        "no-class-assign": "error",
        "no-dupe-class-members": "error",
        "no-duplicate-imports": [
            "error",
            {
                includeExports: true
            }
        ],
        "no-extra-parens": [
            "error",
            "all",
            {
                nestedBinaryExpressions: false
            }
        ],
        "no-inline-comments": "error",
        "no-lonely-if": "error",
        "no-new-symbol": "error",
        "no-mixed-operators": [
            "error",
            {
                allowSamePrecedence: true
            }
        ],
        "no-mixed-spaces-and-tabs": "error",
        "no-multi-assign": "error",
        "no-multiple-empty-lines": "error",
        "no-nested-ternary": "error",
        "no-new-object": "error",
        "no-restricted-syntax": [
            "error",
            "WithStatement"
        ],
        "no-tabs": "error",
        "no-trailing-spaces": "error",
        "no-this-before-super": "error",
        "no-underscore-dangle": "error",
        "no-unneeded-ternary": "error",
        "no-useless-computed-key": "error",
        "no-useless-constructor": "error",
        "no-useless-rename": "error",
        "no-var": "error",
        "no-whitespace-before-property": "error",
        "nonblock-statement-body-position": "error",
        "object-curly-newline": [
            "error",
            {
                multiline: true,
                minProperties: 2
            }
        ],
        "object-curly-spacing": [
            "error",
            "always"
        ],
        "object-property-newline": [
            "error",
            {
                "allowMultiplePropertiesPerLine": false
            }
        ],
        "object-shorthand": [
            "error",
            "always",
            {
                avoidQuotes: true,
                ignoreConstructors: true,
                avoidExplicitReturnArrows: true
            }
        ],
        "one-var": [
            "error",
            "never"
        ],
        "one-var-declaration-per-line": [
            "error",
            'always'
        ],
        "operator-assignment": "error",
        "operator-linebreak": [
            "error",
            "before"
        ],
        "padded-blocks": [
            "error",
            {
                blocks: "never",
                classes: "always",
                switches: "never"
            },
            {
                allowSingleLineBlocks: false
            }
        ],
        "prefer-arrow-callback": [
            "error",
            {
                allowNamedFunctions: true,
                allowUnboundThis: false
            }
        ],
        "prefer-const": "error",
        "prefer-numeric-literals": "error",
        "prefer-object-spread": "error",
        "prefer-rest-params": "error",
        "prefer-spread": "error",
        "prefer-template": "error",
        "quote-props": [
            "error",
            "as-needed"
        ],
        quotes: [
            "error",
            "double",
            {
                avoidEscape: true,
                allowTemplateLiterals: true
            }
        ],
        "require-yield": "error",
        "rest-spread-spacing": "error",
        semi: "error",
        "spaced-comment": [
            "error",
            "always",
            {
                markers: [
                    "!"
                ]
            }
        ],
        "semi-style": "error",
        "sort-keys": [
            "error",
            "asc",
            {
                caseSensitive: false,
                natural: true
            }
        ],
        "sort-imports": "error",
        "space-before-blocks": "error",
        "space-before-function-paren": [
            "error",
            {
                anonymous: "always",
                named: "never",
                asyncArrow: "always"
            }
        ],
        "space-in-parens": [
            "error",
            "never"
        ],
        "space-infix-ops": "error",
        "space-unary-ops": "error",
        "switch-colon-spacing": "error",
        "symbol-description": "error",
        "template-curly-spacing": [
            "error",
            "never"
        ],
        "template-tag-spacing": "error",
        "unicode-bom": "error",
        "yield-star-spacing": [
            "error",
            "after"
        ]
    }
};

export default config;
